crates/dbsp/src/operator/sample.rs

```diff
@@ -0,0 +1,222 @@
+//! Compute random samples of data.
+
+use crate::{
+    algebra::{HasOne, ZRingValue},
+    circuit::{
+        operator_traits::{BinaryOperator, Operator},
+        Scope,
+    },
+    trace::{Batch, BatchReader, Builder},
+    Circuit, DBData, DBWeight, OrdZSet, RootCircuit, Stream,
+};
+use rand::thread_rng;
+use std::{borrow::Cow, marker::PhantomData};
+
+// TODO: Operator to randomly sample `(K, V)` pairs.  This is
+// a little more tricky and also more expensive to implement than
+// sampling keys, as at the low level we need to pick random
+// values (`V`) from `OrderedLayer` and map the back to key (`K`)
+// by doing a lookup in the offset array.
+
+impl<B> Stream<RootCircuit, B>
+where
+    B: BatchReader<Time = ()> + Clone,
+    B::R: ZRingValue,
+{
+    /// Generates a uniform random sample of keys from `self`.
+    ///
+    /// At every clock tick, computes a random sample of the input batch
+    /// using [`BatchReader::sample_keys`].  The `sample_size` stream
+    /// specifies the size of the sample to compute (use `0` when no sample
+    /// is needed at the current clock cycle to make sure the operator
+    /// doesn't waste CPU cycles).
+    ///
+    /// Outputs a Z-set containing randomly sampled keys with weight `1`.
+    ///
+    /// This is not an incremental operator.  It samples the input
+    /// batch received at the current clock cycle and not the integral
+    /// of the input stream.  Prefix the call to `sample_keys()` with
+    /// `integrate_trace()` to sample the integral of the input.
+    ///
+    /// WARNING: This operator (by definition) returns non-dererministic
+    /// outputs.  As such it may not play well with most other DBSP operators
+    /// and must be used with care.
+    pub fn sample_keys(
+        &self,
+        sample_size: &Stream<RootCircuit, usize>,
+    ) -> Stream<RootCircuit, OrdZSet<B::Key, B::R>> {
+        self.circuit().region("sample_keys", || {
+            let stream = self.try_sharded_version();
+
+            // Sample each worker.
+            let local_output = self.circuit().add_binary_operator(
+                SampleKeys::new(),
+                &stream,
+                sample_size,
+            );
+
+            // Sample results from all workers.
+            self.circuit().add_binary_operator(
+                SampleKeys::new(),
+                &local_output.gather(0),
+                sample_size,
+            )
```

Comment:
Why collect all the data on one single thread, why not have each shard select uniform keys and then gather those results? This moves a lot of data to one thread only for that data to be filtered which is really wasteful

Reply:
We start with computing a local sample in each thread and then send locally computed samples to one thread to compute the final result. How would you improve this?

Reply:
Ah never mind, I misunderstood the code